HIP-3.md

```diff
@@ -0,0 +1,223 @@
+     * @dev The `_handleGas` should cover the gas required by the recipient's `handle` function.
+     *   The gas required for ISM costs, including ISM metadata calldata and ISM.verify gas,
+     *   is read from storage and is added to `_handleGas` before being passed to the feeQuotingIgp.
+     * @param _messageId The message ID.
+     * @param _destinationDomain The destination domain of the message.
+     * @param _ismAndHandleGas The amount of gas to cover ISM metadata calldata,
+     *   ISM.verify gas, and gas required by the recipient's `handle` function.
+     * @param _refundAddress The address to refund any msg.value overpayment to.
+     */
+    function payForGas(
+        bytes32 _messageId,
+        uint32 _destinationDomain,
+        uint256 _handleGas,
+        address _refundAddress
+    ) external payable {
+        uint256 _ismAndHandleGas = ismDestinationGas[_destinationDomain] + _handleGas;
+
+        feeQuotingIgp.payForGas(
+            _messageId,
+            _destinationDomain,
+            _ismAndHandleGas,
+            _refundAddress
+        );
+    }
+
+    function setIsmDestinationGas(uint32 _destinationDomain, uint256 _ismGas) external onlyOwner;
+}
+```
+
+The off-chain relayer relating to a `FeeQuotingIgp` will index `GasPayment` events and relate them to dispatched messages based off their message ID. Because the `FeeQuotingIgp` will revert if there was insufficient payment, all `GasPayment` events relate to messages that have fully paid for an amount of gas, and the relayer will honor these payments.
+
+Note that calldata costs for processing the message are not explicitly calculated by the fee quoting IGP. The maximum Hyperlane message size is 2 KiB, or 2048 bytes. Currently, the Ethereum calldata gas cost is 16 gas per non-zero byte, which comes out to a maximum of `32,768` gas. To avoid an additional required mapping in the InterchainGasCalculator of remote domain to the cost per byte of calldata, a flat cost of 33,000 gas to cover potential calldata costs is charged.
+
+### **Rationale**
+
+#### ISM-specific fee quoting IGPs
+
+On the origin chain, an IGP cannot know what the length of an arbitrary ISM's metadata calldata is, nor can it know the expected gas used by an arbitrary ISM's `verify` function. Therefore, the IGP must get this information from some source.
+
+Hyperlane users sending messages that use the default ISM should not need to worry about the destination gas costs incurred by the default ISM. This includes the metadata calldata and the `verify` function. Because the default ISM is out of user control and the parameters of the default ISM may change, a straightforward “happy path” for default ISM users is provided via the default ISM's IGP.
+
+Any non-default ISMs may also have their own ISM-specific IGPs created. Alternatively, users can always call the fee quoting IGP directly with their own `_ismAndHandleGas` amount.
```

Comment:
> Any non-default ISMs may also have their own ISM-specific IGPs created

I still strongly prefer that applications just include Hyperlane overhead in the amount of gas that they pay for

Reply:
By Hyperlane overhead you mean ISM costs right? Fwiw, they can always opt in to doing that by calling the fee quoting IGP directly.

Imo so much of the value of smart contracts is that you're generally able to just deploy something and it just "works" without worrying about operations / things no longer working. Most apps will just use the default ISM (at least for now), and if that default ISM ever changes, app contracts shouldn't be expected to make any changes on their side. Otherwise every time we want to change the quorum threshold, we risk breaking contracts because they could be specifying too low of a gas amount. If we tell people to use a generous buffer of like ~200k and say that the default ISM costs will never exceed that, it still feels like we're constraining what ISMs can do unnecessarily, while overcharging users.

I guess I see this whole ISM-specific IGP thing as providing the simplest experience for users while giving us the most flexibility for future changes without disruption. I'm not sure I understand the benefits of applications including the Hyperlane/ISM overhead themselves?

Reply:
Also fwiw, I feel most strongly about this for the default ISM case where people don't want to have to think about ISMs at all and just want things to work. For more complicated use cases like custom ISMs etc that people explicitly opt into, I feel less strongly